### PR TITLE
Prepopulate email on registration from external provider AB#10923

### PR DESCRIPTION
**Pull request by MikeLyttle**
# Implements [AB#10923](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/10923)

-   [x] Enhancement
-   [ ] Bug
-   [ ] Documentation

## Description

Adds the IdentityProviderEmail field to UserProfileModel, populating it with the email address stored in the JWT.  The registration page on the front-end now prepopulates the email fields with this value when it is present.

If you are reviewing this PR, please adhere to the guidelines as [documented](https://github.com/bcgov/healthgateway/wiki/prguidance).

## Testing

-   [x] Unit Tests Updated
-   [ ] Functional Tests Updated
-   [ ] Not Required

### Steps to Reproduce

If this is a bug, please provide details on how to reproduce the code.

### UI Changes

NO

### Browsers Tested

-   [x] Chrome
-   [ ] Safari
-   [ ] Edge
-   [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments

Any additional notes or comments that may be relevant. Include any specialized deployment steps here.

## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)

-   Fulfills Work Item Requirements

    -   The changes meet/implement story's acceptance criteria. Fixes identified problems if bug.
    -   Changes not directly related to the task requirements need to be documented on the PR.

-   Compilation / Tests

    -   The changes do not break compilation and/or unit or functional tests; unless other item addresses them specifically.

-   Logic Problems / Functional Behaviour

    -   The changes work as intended and do not introduce problems.

-   Performance

    -   The changes do not introduce obvious performance issues.

-   Documented Standards

    -   The changes adhere to the team's documented [Coding Standards](https://github.com/bcgov/healthgateway/wiki/standards).

-   Readability / Maintainability
    -   The changes can be easily understood/read and allow for future enhancements without major refactoring. Readability is preferred over "clever code".
    -   Reasoning for changes needs to be clearly articulated. Disagreements should be arbitrated by a third developer.

### CODE REVIEW

**Review by sslaws — APPROVED** [review]

File: Apps/WebClient/src/ClientApp/src/models/userProfile.ts [hunk 1]
```diff
@@ -15,6 +15,9 @@ export default interface UserProfile {
     // Indicates whether the email was verified.
     isEmailVerified: boolean;
 
+    // The email from the user's external identity provider.
+    identityProviderEmail?: string | null;
```
Comment (FuriousLlama) [inline Apps/WebClient/src/ClientApp/src/models/userProfile.ts:19]: The question mark at the end implies it could be undefined. I am guessing is only going to be a string or null right?

Comment (MikeLyttle) [reply Apps/WebClient/src/ClientApp/src/models/userProfile.ts:19]: I think this is the model that's also used when we save changes on the profile page.  In that case, we don't specify a value for the field.

File: Apps/WebClient/src/Server/Services/UserProfileService.cs [hunk 2]
```diff
@@ -109,7 +109,7 @@ public RequestResult<UserProfileModel> GetUserProfile(string hdid, DateTime jwtA
                 return new RequestResult<UserProfileModel>()
                 {
                     ResultStatus = ResultType.Success,
-                    ResourcePayload = new UserProfileModel(),
+                    ResourcePayload = new UserProfileModel() { IdentityProviderEmail = jwtEmailAddress },
```
Comment (FuriousLlama) [inline Apps/WebClient/src/Server/Services/UserProfileService.cs:112]: this is odd. Why is it necessary to set the jwt email to the one passed? If there are no profiles on the backend the object is empty indicating that

File: Apps/WebClient/src/Server/Services/UserProfileService.cs [hunk 3]
```diff
@@ -125,6 +125,7 @@ public RequestResult<UserProfileModel> GetUserProfile(string hdid, DateTime jwtA
             RequestResult<TermsOfServiceModel> termsOfServiceResult = this.GetActiveTermsOfService();
 
             UserProfileModel userProfile = UserProfileModel.CreateFromDbModel(retVal.Payload);
+            userProfile.IdentityProviderEmail = jwtEmailAddress;
```
Comment (FuriousLlama) [inline Apps/WebClient/src/Server/Services/UserProfileService.cs:128]: why is this necessary? Wouldnt both be the same?

Comment (MikeLyttle) [reply Apps/WebClient/src/Server/Services/UserProfileService.cs:128]: This is where IdentityProviderEmail is populated.  `jwtEmailAddress` is fetched in GetUserProfile() in the controller.

File: Apps/WebClient/src/Server/Services/UserProfileService.cs (hunk 2 above)
Comment (MikeLyttle) [reply Apps/WebClient/src/Server/Services/UserProfileService.cs:112]: This is where IdentityProviderEmail is populated.  `jwtEmailAddress` is fetched in GetUserProfile() in the controller.

File: Apps/WebClient/src/Server/Controllers/UserProfileController.cs [hunk 4]
```diff
@@ -126,8 +126,9 @@ public IActionResult GetUserProfile(string hdid)
             this.logger.LogTrace($"HTTP context user: {JsonConvert.SerializeObject(user, jsonSettings)}");
 
             DateTime jwtAuthTime = GetAuthDateTime(user);
+            string? jwtEmailAddress = user.FindFirstValue(ClaimTypes.Email);
 
-            RequestResult<UserProfileModel> result = this.userProfileService.GetUserProfile(hdid, jwtAuthTime);
+            RequestResult<UserProfileModel> result = this.userProfileService.GetUserProfile(hdid, jwtAuthTime, jwtEmailAddress);
```
Comment (FuriousLlama) [inline Apps/WebClient/src/Server/Controllers/UserProfileController.cs:131]: I am not sure the email from the jwt should be retrieved if it's not the one the user created. Are we thinking about displaying both somewhere?

Comment (MikeLyttle) [reply Apps/WebClient/src/Server/Controllers/UserProfileController.cs:131]: Let's discuss.

File: Apps/WebClient/src/ClientApp/src/views/registration.vue [hunk 5]
```diff
@@ -71,7 +75,10 @@ export default class RegistrationView extends Vue {
         this.logger = container.get<ILogger>(SERVICE_IDENTIFIER.Logger);
         this.minimumAge = this.webClientConfig.minPatientAge;
 
-        if (
+        if (this.user.identityProviderEmail !== null) {
+            this.email = this.user.identityProviderEmail;
+            this.emailConfirmation = this.user.identityProviderEmail;
```
Comment (FuriousLlama) [inline Apps/WebClient/src/ClientApp/src/views/registration.vue:80]: This looks odd. The user has not been created yet if they are on the registration page.

File: Apps/WebClient/src/ClientApp/src/models/userProfile.ts (hunk 1 above)
Comment (FuriousLlama) [reply Apps/WebClient/src/ClientApp/src/models/userProfile.ts:19]: Added some more questions related to this field. It does not look correct to have two types of emails on the profile object

Comment (MikeLyttle) [reply Apps/WebClient/src/ClientApp/src/models/userProfile.ts:19]: We should discuss.

File: Apps/WebClient/src/Server/Services/UserProfileService.cs (hunk 3 above)
Comment (FuriousLlama) [reply Apps/WebClient/src/Server/Services/UserProfileService.cs:128]: AFAIK the JWT is read on the backend to make sure that is validated, all other flows work the same

File: Apps/WebClient/src/ClientApp/src/models/userProfile.ts (hunk 1 above)
Comment (MikeLyttle) [reply Apps/WebClient/src/ClientApp/src/models/userProfile.ts:19]: Will swap to retrieving the email from the JWT on the front-end for a cleaner implementation.

**Review by tiagograf-AOT — APPROVED** [review]

**Review by FuriousLlama — APPROVED** [review]

File: Apps/WebClient/src/ClientApp/src/views/registration.vue [hunk 6]
```diff
@@ -96,6 +96,11 @@ export default class RegistrationView extends Vue {
                 if (oidcUser) {
                     this.oidcUser = oidcUser;
 
+                    if (this.oidcUser.email !== null) {
+                        this.email = this.oidcUser.email;
+                        this.emailConfirmation = this.oidcUser.email;
```
Comment (FuriousLlama) [inline Apps/WebClient/src/ClientApp/src/views/registration.vue:101]: Much cleaner
